keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,645 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kustomize provides an intuitive way to manage Kubernetes applications in a
+purely declarative manner. However, it seems to struggle with applications that
+mix multiple, optional features on demand, affecting different aspects of a
+base configuration.
+
+This enhancement proposal introduces `Components` a new type of Kustomization
+that allows users to define reusable kustomizations and then
+include them from higher-level overlays based on their needs.
+
+The community has shown strong interest in this feature and has been actively
+discussing it in https://github.com/kubernetes-sigs/kustomize/issues/1251.
+
+## Motivation
+
+By design, the variant system of Kustomize is based on an inheritance model
+and suggests that a base application is parameterized by stacking overlays on
+top of it. While intuitive at first sight, this approach falls short of
+deploying multivariate applications, i.e., ones that:
+
+- ship with numerous opt-in features and corresponding configuration options
+- target dissimilar audiences with different needs and scopes (developers,
+  admins, customers, testers etc.)
+- run on multiple cloud platforms (on-prem, Minikube, GKE, EKS, AKS, etc.)
+
+The problem is that modular applications cannot always be expressed in a tall
+hierarchy while preserving all combinations of available features. Doing so
+would require putting each feature in an overlay, and making overlays for
+independent features inherit from each other. However, this is semantically
+incorrect, cannot not scale as the number of features grows, and soon results
+in duplicate manifests and kustomizations.
+
+Instead, such applications are much better expressed as a collection of
+components, i.e., reusable pieces of configuration logic that are defined in a
+common place and that distinct overlays can then mix-and-match. This approach
+abides by the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
+principle and increases ease of maintenance.
+
+The simplest way to implement this in Kustomize is to create an overlay for
+each component, and have a top-level overlay include only the components that
+it requires. The problem with this approach is that the mechanism that
+Kustomize uses to interpret and ultimately apply patches does not allow for
+mutating different aspects of the very same base resource in parallel without
+modifying its GVKN (unique identifier). That is, sibling overlays that operate
+on the same level of the hierarchy and are not chained (i.e., one does not
+import the other as base) cannot modify their common parent due to resource ID
+conflicts.
+
```

Comment:
Here we can point to existing issues that Kustomize users have raised and are relevant with this KEP: 

For example:

- https://github.com/kubernetes-sigs/kustomize/issues/171
- https://github.com/kubernetes-sigs/kustomize/issues/759
- https://github.com/kubeflow/kubeflow/issues/3490#issuecomment-571365617
https://github.com/kubeflow/kubeflow/issues/3490#issuecomment-579395304

Reply:
I would substitute the Kubeflow comments with this PR: https://github.com/kubeflow/kubeflow/pull/3108

Also, what about these issues?

* https://github.com/kubernetes-sigs/kustomize/issues/942
* https://github.com/kubernetes-sigs/kustomize/issues/2464

Reply:
@apyrgio I saw you had noticed the issue I made :) My naive idea for inheritance would have been something like:

```
abstract:
  apiVersion: batch/v1beta1
  kind: CronJob
  metadata:
    name: test-cron1
  spec:
    concurrencyPolicy: Forbid
    jobTemplate:
      spec:
        parallelism: 1
        completions: 1
        backoffLimit: 2
        template:
          spec:
            nodeSelector:
              highmem: "true"
            tolerations:
              - key: reserved-pool
                operator: Equal
                value: "true"
                effect: NoSchedule
            containers:
              - name: test-cron
                image: gcr.io/proc/app:latest
                resources:
                  requests:
                    memory: "4Gi"
            restartPolicy: Never
```

and usage:

```
apiVersion: batch/v1beta1
inherits: ./base.yaml # the above example basically
spec:
  schedule: '0 3 * * *'
  jobTemplate:
    spec:
      template:
        spec:
          containers:
              command:
                [
                  'node',
                  '--max-old-space-size=8192',
                  '--unhandled-rejections=strict',
                  'test.js',
                ]
```

Where `abstract` files could be partial, as well as `inherits` files. This would be very IDE friendly: IDE could understand that `abstract` can have missing fields and is not supposed to be applied as is, and it could also understand that `inherits` needs to have all the required fields filled and that some of them comes from the abstract file. This would help with auto complete and incorrect file detection etc.

Just my two cents :)